### ISSUE TITLE
CMake 3.31.0 causes breakage with cmake module due to `--dependency-file`

### ISSUE THREAD

**Issue by pobrn** [mesonbuild/meson#13888]
**Describe the bug**

Upgrading from cmake 3.30.5 to 3.31.0 causes linking not to succeed when a cmake subproject is used.

**To Reproduce**

```meson
# meson.build

project(
	'test',
	['cpp'],
	default_options: [
		'cpp_std=c++17',
	],
)

cmake = import('cmake')
libyuv_vars = cmake.subproject_options()
libyuv_vars.set_override_option('cpp_std', 'c++17')
libyuv_vars.append_link_args('-ljpeg')
libyuv = cmake.subproject('libyuv', options : libyuv_vars)
libyuv_dep = libyuv.dependency('yuv')

executable(
	'main',
	'main.cpp',
	dependencies: [ libyuv_dep ],
)
```

```ini
# subprojects/libyuv.wrap

[wrap-git]
directory = libyuv
url = https://chromium.googlesource.com/libyuv/libyuv.git
revision = 93b1b332cd60b56ab90aea14182755e379c28a80
```

Then simply run

```
$ meson setup build
$ meson compile -C build
```

And observe a linker error:

```
ninja: Entering directory `/tmp/asd/build'
[102/102] Linking target subprojects/libyuv/libyuv_shared.so
FAILED: subprojects/libyuv/libyuv_shared.so 
c++  -o subprojects/libyuv/libyuv_shared.so subprojects/libyuv/libyuv_shared.so.p/source_compare.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_from.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_from_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_jpeg.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_to_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_to_i420.cc.o subprojects/libyuv/libyuv_shared.so.p/source_cpu_id.cc.o subprojects/libyuv/libyuv_shared.so.p/source_mjpeg_decoder.cc.o subprojects/libyuv/libyuv_shared.so.p/source_mjpeg_validate.cc.o subprojects/libyuv/libyuv_shared.so.p/source_planar_functions.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_any.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_any.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_any.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_uv.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_video_common.cc.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,libyuv_shared.so -Wl,--dependency-file,CMakeFiles/yuv_shared.dir/link.d -ljpeg
/usr/bin/ld: cannot open dependency file CMakeFiles/yuv_shared.dir/link.d: no error
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Cause**

<details>
<summary>The difference between the `build.ninja` files</summary>

```diff
--- build.ninja-3.30.5	2024-11-09 20:26:09.473873115 +0100
+++ build.ninja-3.31.0	2024-11-09 20:27:48.973123121 +0100
@@ -544,7 +544,7 @@
  IMPLIB = subprojects/libyuv/libyuv_shared.so
 
 build subprojects/libyuv/libyuv_shared.so: cpp_LINKER subprojects/libyuv/libyuv_shared.so.p/source_compare.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_compare_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_from.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_from_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_jpeg.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_to_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_convert_to_i420.cc.o subprojects/libyuv/libyuv_shared.so.p/source_cpu_id.cc.o subprojects/libyuv/libyuv_shared.so.p/source_mjpeg_decoder.cc.o subprojects/libyuv/libyuv_shared.so.p/source_mjpeg_validate.cc.o subprojects/libyuv/libyuv_shared.so.p/source_planar_functions.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_any.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_rotate_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_any.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_row_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_any.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_argb.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_common.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_gcc.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_mmi.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_msa.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_neon.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_neon64.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_uv.cc.o subprojects/libyuv/libyuv_shared.so.p/source_scale_win.cc.o subprojects/libyuv/libyuv_shared.so.p/source_video_common.cc.o | /usr/lib/libjpeg.so.8.3.2
- LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,libyuv_shared.so -ljpeg
+ LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,libyuv_shared.so -Wl,--dependency-file,CMakeFiles/yuv_shared.dir/link.d -ljpeg
 
 build subprojects/libyuv/yuvconvert.p/util_yuvconvert.cc.o: cpp_COMPILER ../subprojects/libyuv/util/yuvconvert.cc
  DEPFILE = subprojects/libyuv/yuvconvert.p/util_yuvconvert.cc.o.d
@@ -552,7 +552,7 @@
  ARGS = -Isubprojects/libyuv/yuvconvert.p -Isubprojects/libyuv -I../subprojects/libyuv -I../subprojects/libyuv/include -Isubprojects/libyuv/__CMake_build -I../subprojects/libyuv/__CMake_build -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++20 -O0 -g -fPIE -g -DHAVE_JPEG -Wno-sign-compare -Wno-unused-variable -Wno-unused-parameter
 
 build subprojects/libyuv/yuvconvert: cpp_LINKER subprojects/libyuv/yuvconvert.p/util_yuvconvert.cc.o | /usr/lib/libjpeg.so.8.3.2 subprojects/libyuv/libyuv.a
- LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -pie -Wl,--start-group subprojects/libyuv/libyuv.a -g -rdynamic -ljpeg -Wl,--end-group
+ LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -pie -Wl,--start-group subprojects/libyuv/libyuv.a -g -rdynamic -Wl,--dependency-file,CMakeFiles/yuvconvert.dir/link.d -ljpeg -Wl,--end-group
 
 build main.p/main.cpp.o: cpp_COMPILER ../main.cpp
  DEPFILE = main.p/main.cpp.o.d
```

</details>

<details>
<summary>The difference between the generated `meson.build` files</summary>

```diff
--- meson.build-3.30.5	2024-11-09 20:49:29.619985896 +0100
+++ meson.build-3.31.0	2024-11-09 20:48:16.870534254 +0100
@@ -131,7 +131,7 @@
   'yuv_shared',
   yuv_shared_src,
   build_by_default : true,
-  link_args : ['-ljpeg'],
+  link_args : ['-Wl,--dependency-file,CMakeFiles/yuv_shared.dir/link.d', '-ljpeg'],
   link_with : [],
   include_directories : yuv_shared_inc,
   install : true,
@@ -154,7 +154,7 @@
   'yuvconvert',
   yuvconvert_src,
   build_by_default : false,
-  link_args : ['-g', '-rdynamic', '-ljpeg', '-ljpeg'],
+  link_args : ['-g', '-rdynamic', '-Wl,--dependency-file,CMakeFiles/yuvconvert.dir/link.d', '-ljpeg', '-ljpeg'],
   link_with : [yuv],
   include_directories : yuvconvert_inc,
   install : false,
```

</details>

<details>
<summary>The difference between the outputs of the cmake invocations</summary>

```diff
--- cmake.out-3.30.5	2024-11-09 20:49:10.973459780 +0100
+++ cmake.out-3.31.0	2024-11-09 20:50:21.346262665 +0100
@@ -1,6 +1,6 @@
-libyuv| Found CMake: /usr/bin/cmake (3.30.5)
+libyuv| Found CMake: /usr/bin/cmake (3.31.0)
 
-| Configuring the build directory with CMake version 3.30.5
+| Configuring the build directory with CMake version 3.31.0
 | Running CMake with: -G Ninja -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Debug
 |   - build directory:          /tmp/asd/build/subprojects/libyuv/__CMake_build
 |   - source directory:         /tmp/asd/subprojects/libyuv
@@ -22,7 +22,7 @@
 
 | -- The C compiler identification is GNU 14.2.1
 | -- The CXX compiler identification is GNU 14.2.1
-| CMake Warning (dev) at /usr/share/cmake/Modules/CMakeFindBinUtils.cmake:235 (find_program):
+| CMake Warning (dev) at /usr/share/cmake/Modules/CMakeFindBinUtils.cmake:225 (find_program):
 | Policy CMP0126 is not set: set(CACHE) does not remove a normal variable of
 | the same name.  Run "cmake --help-policy CMP0126" for policy details.  Use
 | the cmake_policy command to set the policy and suppress this warning.
@@ -45,7 +45,7 @@
 | -- Detecting CXX compile features
 | -- Detecting CXX compile features - done
 | CMake Deprecation Warning at CMakeLists.txt:6 (CMAKE_MINIMUM_REQUIRED):
-| Compatibility with CMake < 3.5 will be removed from a future version of
+| Compatibility with CMake < 3.10 will be removed from a future version of
 | CMake.
 
 | Update the VERSION argument <min> value or use a ...<max> suffix to tell
@@ -64,7 +64,7 @@
 
 | Building ver.: 0.0.1770
 | Packaging for: amd-64
-| -- Configuring done (4.1s)
+| -- Configuring done (1.5s)
 | CMake Warning (dev) at CMakeLists.txt:26 (ADD_LIBRARY):
 | Policy CMP0156 is not set: De-duplicate libraries on link lines based on
 | linker capabilities.  Run "cmake --help-policy CMP0156" for policy details.
@@ -92,5 +92,5 @@
 | be applied.
 | This warning is for project developers.  Use -Wno-dev to suppress it.
 
-| -- Generating done (0.1s)
+| -- Generating done (0.0s)
 | -- Build files have been written to: /tmp/asd/build/subprojects/libyuv/__CMake_build
```

</details>

<details>
<summary>cmake biset</summary>

```
git bisect start
# status: waiting for both good and bad commits
# bad: [23e24788abb58fb7ac3e6ce71fa7fd59dbd8f0ef] CMake Nightly Date Stamp
git bisect bad 23e24788abb58fb7ac3e6ce71fa7fd59dbd8f0ef
# status: waiting for good commit(s), bad commit known
# good: [9c4a0a9ff09735b847bbbc38caf6da7f6c7238f2] CMake 3.30.5
git bisect good 9c4a0a9ff09735b847bbbc38caf6da7f6c7238f2
# bad: [006891dab90ea72f634e83858a600aa61a40d004] Tests/RunCMake/BuildDepends: Prepare CompileDepends check for multiple targets
git bisect bad 006891dab90ea72f634e83858a600aa61a40d004
# bad: [5277769a93e74ae299ce3002a480022c445aadf5] Merge branch 'release-3.28'
git bisect bad 5277769a93e74ae299ce3002a480022c445aadf5
# good: [ddcba76547a0a2fce7422e2cccce40c51708a600] Merge topic 'bash-completion-cmake-workflow'
git bisect good ddcba76547a0a2fce7422e2cccce40c51708a600
# good: [62bb8a77a590c705b73ea80783d2d2693e70cfa8] cmMakefile: Teach ConfigureFile to move the temp file instead of copying it
git bisect good 62bb8a77a590c705b73ea80783d2d2693e70cfa8
# bad: [70b4966c01ced74455cd38015fe2294fd762d6c0] Merge topic 'refactor-lexer'
git bisect bad 70b4966c01ced74455cd38015fe2294fd762d6c0
# good: [bb497152f08d3dff593132ea2ba1e0dbd70952a5] Merge topic 'llpkgc-integration'
git bisect good bb497152f08d3dff593132ea2ba1e0dbd70952a5
# good: [8a833fbf9b04779b7339cee1d60bd59ac0f047bc] Merge branch 'release-3.30'
git bisect good 8a833fbf9b04779b7339cee1d60bd59ac0f047bc
# bad: [e27a050bff327ef663465705fedc7e033a4a8b94] Merge topic 'linker-configuration'
git bisect bad e27a050bff327ef663465705fedc7e033a4a8b94
# good: [7d9c4dfd0bcd8b05d438b255d69942341a918e86] Merge branch 'release-3.30'
git bisect good 7d9c4dfd0bcd8b05d438b255d69942341a918e86
# bad: [1e35163ae8a333f7586aaa7c4f3a754a5018db49] WHOLE_ARCHIVE link feature: rely now on linker configuration
git bisect bad 1e35163ae8a333f7586aaa7c4f3a754a5018db49
# bad: [c1c4cf9545d17928c065fca344c48e03b80291f0] Linker configuration: introduce a new architecture
git bisect bad c1c4cf9545d17928c065fca344c48e03b80291f0
# first bad commit: [c1c4cf9545d17928c065fca344c48e03b80291f0] Linker configuration: introduce a new architecture
```

</details>

Bisecting cmake points to [commit c1c4cf9545d17928c065fca344c48e03b80291f0](https://gitlab.kitware.com/cmake/cmake/-/commit/c1c4cf9545d17928c065fca344c48e03b80291f0) added in https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9665. This change touches many files, but it seems that change did not introduce the use of `--dependency-file`, so maybe there was an issue with cmake earlier and it did not actually use that argument.

I am wondering if maybe it is enough to strip the `--dependency-file` linker options when it specifies a path inside `CMakeFiles` and fix the path for manually specified `--dependency-file`.

**Workaround**

Add `CMAKE_LINK_DEPENDS_USE_LINKER=FALSE`, e.g.:
```meson
libyuv_vars.add_cmake_defines({'CMAKE_LINK_DEPENDS_USE_LINKER': 'FALSE'})
```

This is another thing that meson's cmake module could potentially do.

**system parameters**
* native build
* python: `3.12.7`
* meson: `1.6.0`
* ninja: `1.12.1`

**Comment by PHANTOM-2004** [comment]
I met similar issue too

**Comment by DimStar77** [comment]
meson also fails its own test suite with cmake 3.31.0

```
[  166s] [1/3] Generating subprojects/cmMod/custom_tgt_0 with a custom command
[  166s] /home/abuild/rpmbuild/BUILD/meson-1.6.0/test cases/cmake/11 cmake_module_path/subprojects/cmMod/gen.py:1: EncodingWarning: 'encoding' argument not specified
[  166s]   with open('main.c', 'w') as fp:
[  166s] [2/3] Compiling C object subprojects/cmMod/main.p/meson-generated_.._main.c.o
[  166s] [3/3] Linking target subprojects/cmMod/main
[  166s] FAILED: subprojects/cmMod/main 
[  166s] cc  -o subprojects/cmMod/main subprojects/cmMod/main.p/meson-generated_.._main.c.o -Wl,--as-needed -Wl,--no-undefined -g -Wl,--dependency-file,CMakeFiles/main.dir/link.d
[  166s] /usr/lib64/gcc/x86_64-suse-linux/14/../../../../x86_64-suse-linux/bin/ld: cannot open dependency file CMakeFiles/main.dir/link.d: no error
[  166s] collect2: error: ld returned 1 exit status
[  166s] ninja: build stopped: subcommand failed.
```

**Comment by thesamesam** [comment]
Just to check here -- what version of binutils is this? I'm wondering if you're hitting https://gitlab.kitware.com/cmake/cmake/-/issues/26401 (https://sourceware.org/bugzilla/show_bug.cgi?id=32300) or if that's a totally different issue.

**Comment by heftig** [comment]
This change makes the test suite pass again:

```
diff --git c/mesonbuild/cmake/interpreter.py i/mesonbuild/cmake/interpreter.py
index fafee86ab..12d7049d2 100644
--- c/mesonbuild/cmake/interpreter.py
+++ i/mesonbuild/cmake/interpreter.py
@@ -828,6 +828,7 @@ class CMakeInterpreter:
 
         cmake_args = []
         cmake_args += cmake_get_generator_args(self.env)
+        cmake_args += ['-DCMAKE_LINK_DEPENDS_USE_LINKER=FALSE']
         cmake_args += [f'-DCMAKE_INSTALL_PREFIX={self.install_prefix}']
         cmake_args += extra_cmake_options
         if not any(arg.startswith('-DCMAKE_BUILD_TYPE=') for arg in cmake_args):
```

But I'm not sure it's right.

**Comment by Habbie** [comment]
If you happen to be running into this because Github upgraded CMake in a runner image, this works for me: https://github.com/voorkant/voorkant-core/pull/140/files

**Comment by eli-schwartz** [comment]
Suppressing it seems plausible since it is implementing generic target functionality that, if meson wants it, should be handled by the meson backend. It really has nothing to do with target generation.

**Comment by eli-schwartz** [comment]
It was reported 3 days ago at https://gitlab.kitware.com/cmake/cmake/-/issues/26498 and now fixed in cmake 3.31.2 as a regression. To quote cmake devs:

> [@marc.chevrier](https://github.com/marc.chevrier) I think the problem is that we're now exposing the `-Wl,--dependency-file,...` flag in the file-api's reply.  The linker dependencies flag is an implementation detail that is not specified by project code, and therefore shouldn't be reported in the file-api reply.  Similarly, we don't report compiler depfile flags.

> [!10061 (merged)](https://github.com/cmake/cmake/-/merge_requests/10061) drops the `--dependency-file` from the file-api reply.  It's actually been there since CMake 3.27, but went unnoticed before [!9665 (merged)](https://github.com/cmake/cmake/-/merge_requests/9665) enabled linker-generated dependencies more broadly.
